### PR TITLE
feat(cli): replace Docker-based get-sync-token with native Obsidian API call

### PR DESCRIPTION
**Pull request by aliasunder**
## Summary

- Replace the Docker-based `get-sync-token` with a direct `fetch` to `api.obsidian.md/user/signin` — the same endpoint the `obsidian-headless` CLI uses. Removes the 2 GB Docker Desktop prerequisite from token capture.
- The new flow prompts for email, password, and MFA code (when 2FA is enabled) via clack prompts, with a 30s timeout, spinner feedback, and per-failure-mode error messages.
- `init --mode remote` now always offers "Generate the token now?" regardless of Docker status — the token capture no longer needs Docker at all.
- Remove `runObsidianLogin` from `DockerRunner`, `buildObsidianLoginArgs`, and all test stubs — dead code after this change.
- Update CLI README, deploy guides (Railway, Render, remote), and the `get-sync-token` program description.

## Design

The API contract was verified from `obsidian-headless` v0.0.14 source:

- `POST https://api.obsidian.md/user/signin` with `{ email, password, mfa }` and `Origin: https://obsidian.md` header
- Success: `{ token, name, email }` — no `error` field
- MFA required: 200 with `{ error: "...2FA code..." }` — prompt and retry
- Wrong MFA: `{ error: "...2FA code is incorrect..." }` — no retry
- Wrong password: `{ error: "Invalid email or password" }`
- Server error: HTTP non-200

The MFA detection mirrors the exact logic from the `ob` CLI's login command.

`fetchFn: typeof fetch` is injected via DI (same pattern as `init.ts`'s existing `fetchFn`), making the API call fully testable without mocking globals.

## Test plan

- [x] 15 unit tests for `captureObsidianToken` and `runGetSyncToken` — success, MFA flow, wrong MFA, wrong password, HTTP 500, HTTP 429, network error, timeout, non-JSON response, missing token field, --dir write, --dir missing .env
- [x] 10 init tests updated — confirm always offered, auto-capture uses mock fetch, fallback uses mock fetch error
- [x] 3 `buildObsidianLoginArgs` tests removed (dead code)
- [x] Full suite: 3,063 tests pass, zero lint errors, clean build
- [ ] Manual: `npx vault-cortex@latest get-sync-token` with no Docker installed (beta publish)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Sync tokens can now be captured directly through the Obsidian API without Docker.
  * Sign-in supports email, password, optional MFA, retries, and manual token entry as a fallback.
  * Remote initialization offers automatic token capture regardless of Docker availability.

* **Documentation**
  * Deployment guides now recommend Node.js and `npx` for token setup, with Docker as an optional fallback.
  * CLI descriptions now reflect the direct account sign-in process.

### CODE REVIEW

File: cli/src/get-sync-token.ts [hunk 1]
```diff
@@ -17,173 +10,169 @@ export type GetSyncTokenFlags = {
...
+        spinner.stop(`Signed in as ${result.name} (${result.email}).`)
+        return result.token
+      } catch (retryError) {
+        spinner.stop("Sign-in failed.")
+        prompts.warn(
+          `Could not sign in: ${describeError(retryError)}\n` +
+            "  Check your 2FA code and try again.",
+        )
+        return undefined
+      }
```
[the remark this answers is not in the document]
Comment (aliasunder) [reply cli/src/get-sync-token.ts]: Fixed in 7ce7b17 — the retry catch now routes TimeoutError to the timeout message and only appends the 2FA hint for ObsidianApiError responses. Network errors on retry get a plain message with no misleading code-check guidance. Two new tests cover the retry-timeout and retry-network-error paths.

---
*🔍 ship-check · pr-monitor · Claude Opus 4.6 (1M context)*

File: cli/src/__tests__/get-sync-token.test.ts [hunk 2]
```diff
@@ -1,254 +1,281 @@
...
+      fetchFn: fetchApiError("2FA code is incorrect"),
+    })
 
     expect(token).toBeUndefined()
     expect(scripted.warnings[0]).toBe(
-      "The Obsidian login did not complete — you can run it later with:\n" +
-        "  npx vault-cortex@latest get-sync-token",
+      "Could not sign in: 2FA code is incorrect",
     )
   })
```
[the remark this answers is not in the document]
Comment (aliasunder) [reply cli/src/__tests__/get-sync-token.test.ts:165]: Already covered — Phase 3 (test audit, d920916) added this exact test: `fetchMfaRetryFail("2FA code is incorrect")` exercises the post-MFA-prompt retry path and asserts the specific warning with the 2FA hint. The bot reviewed at 2a661e3 before that commit landed.

---
*🔍 ship-check · pr-monitor · Claude Opus 4.6 (1M context)*

**Comment by aliasunder** [comment]
**Re: umm-actually AGENTS.md finding** — already fixed in d939c87 (Phase 1). The structure tree now reads "Obsidian Sync token capture via the Obsidian signin API". The bot reviewed at 2a661e3 before that commit.

---
*🔍 ship-check · pr-monitor · Claude Opus 4.6 (1M context)*

File: cli/src/__tests__/get-sync-token.test.ts [hunk 3]
```diff
@@ -1,254 +1,390 @@
...
+    const token = await captureObsidianToken({
+      prompts: scripted.prompts,
+      fetchFn: fetchNonJson(),
+    })
+
+    expect(token).toBeUndefined()
+    expect(scripted.warnings[0]).toBe(
+      "Could not sign in: Unexpected response from Obsidian API (not JSON)",
     )
+  })
```
[the remark this answers is not in the document]
Comment (aliasunder) [reply cli/src/__tests__/get-sync-token.test.ts:334]: Fixed in d9b07b6 — added a `fetchJsonNonObject` mock (returns `[1, 2, 3]`) and a test that asserts the same "not JSON" warning. Also tightened the `isRecord` guard to exclude arrays (`!Array.isArray(value)`).

---
*🔍 ship-check · pr-monitor · Claude Opus 4.6 (1M context)*

File: cli/src/get-sync-token.ts [hunk 4]
```diff
@@ -17,173 +10,178 @@ export type GetSyncTokenFlags = {
...
+        if (retryError instanceof Error && retryError.name === "TimeoutError") {
+          prompts.warn(
+            "Request timed out — check your internet connection and try again.",
+          )
+          return undefined
+        }
+        const retryHint =
+          retryError instanceof ObsidianApiError
+            ? "\n  Check your 2FA code and try again."
+            : ""
```
[the remark this answers is not in the document]
Comment (aliasunder) [reply cli/src/get-sync-token.ts]: Fixed in d9b07b6 — the retry hint is now gated on `retryError.message.includes("2FA code")` so account-level errors (rate limiting, lockout) get a plain message without the misleading code-check guidance.

---
*🔍 ship-check · pr-monitor · Claude Opus 4.6 (1M context)*

File: deploy/railway/README.md [hunk 5]
```diff
@@ -61,7 +61,7 @@ your computer — once, before you click the button.
 4. **Copy the token.** The deploy form asks for it as `OBSIDIAN_AUTH_TOKEN`.
 
 Already have Node.js? `npx vault-cortex@latest get-sync-token` does the same
-(Docker still has to be running).
+without Docker.
```
[the remark this answers is not in the document]
Comment (aliasunder) [reply deploy/railway/README.md]: Fixed in d9b07b6 — both Railway and Render guides now lead with `npx vault-cortex@latest get-sync-token` as the primary path. Docker is in a collapsible "Don't have Node.js?" fallback. Matches the remote guide's structure.

---
*🔍 ship-check · pr-monitor · Claude Opus 4.6 (1M context)*

**Comment by aliasunder** [comment]
@coderabbit review

File: cli/src/init.ts [hunk 6]
```diff
@@ -445,25 +446,17 @@ const runRemoteInit = async (
...
+  // When the user declines or capture fails, the token is left blank in .env
+  // and get-sync-token can fill it in later.
   const obsidianAuthToken =
-    capturedToken ??
-    (
-      await prompts.password(
-        "Paste the Obsidian Sync token (leave blank to fill in .env later):",
-      )
-    ).trim()
+    (await offerSyncTokenCapture(prompts, fetchFn)) ?? ""
```
[the remark this answers is not in the document]
Comment (aliasunder) [reply cli/src/init.ts]: Fixed in 9022b7c — when capture is declined on a re-init, the flow now reads `OBSIDIAN_AUTH_TOKEN` from the on-disk `.env` (via `readEnvObsidianToken`). An existing token preserves the start offer and skips the "fill in your token" guidance. Added `readEnvObsidianToken` to `scaffold.ts` following the same pattern as `readEnvPublicUrl`.

---
*🔍 ship-check · pr-monitor · Claude Opus 4.6 (1M context)*

File: cli/src/__tests__/get-sync-token.test.ts [hunk 7]
```diff
@@ -1,254 +1,412 @@
...
+        JSON.stringify({ error: "Your account requires a 2FA code" }),
+        { status: 200, headers: { "Content-Type": "application/json" } },
+      )
+    }
+    return new Response(JSON.stringify({ error: retryError }), {
+      status: 200,
+      headers: { "Content-Type": "application/json" },
+    })
+  }) as typeof fetch
+}
```
[the remark this answers is not in the document]
Comment (aliasunder) [reply cli/src/__tests__/get-sync-token.test.ts]: This is the standard mock pattern used throughout the CLI tests (`fetchOk`, `fetchNever` in `command-stubs.ts`). The mock only needs to return a `Response` — adding the full `fetch` parameter signature to every mock factory would be noise. The lint correctly reports it as a warning, not an error.

---
*🔍 ship-check · pr-monitor · Claude Opus 4.6 (1M context)*

File: cli/src/get-sync-token.ts [hunk 8]
```diff
@@ -17,173 +10,174 @@ export type GetSyncTokenFlags = {
...
     )
+    return
   }
+
+  const isMfaError =
+    error instanceof ObsidianApiError && error.message.includes("2FA code")
+  const mfaHint =
+    isMfaRetry && isMfaError ? "\n  Check your 2FA code and try again." : ""
+
+  prompts.warn(`Could not sign in: ${describeError(error)}${mfaHint}`)
```
[the remark this answers is not in the document]
Comment (aliasunder) [reply cli/src/get-sync-token.ts:96]: Already fixed — the retry catch routes TimeoutError separately and gates the 2FA hint on `isMfaError` (an ObsidianApiError whose message contains "2FA code"). See `warnSigninError` at line 78.

---
*🔍 ship-check · pr-monitor · Claude Opus 4.6 (1M context)*

File: AGENTS.md [hunk 9]
```diff
@@ -98,7 +98,7 @@ cli/                                   # npx vault-cortex CLI (published as vaul
     docker.ts                          # Container management (docker run, health-check wait)
     upgrade.ts                         # Upgrade command (pull + re-create + health check)
     lifecycle.ts                       # Down/logs/restart commands + shared deployment resolution and re-create plumbing
-    get-sync-token.ts                  # Get-sync-token subcommand (Sync token auto-capture via volume mount)
+    get-sync-token.ts                  # Get-sync-token subcommand (Sync token capture via Obsidian API)
```
[the remark this answers is not in the document]
Comment (aliasunder) [reply AGENTS.md:101]: Already fixed — the structure tree now reads "Obsidian Sync token capture via the Obsidian signin API" (d939c87).

---
*🔍 ship-check · pr-monitor · Claude Opus 4.6 (1M context)*

File: cli/src/__tests__/get-sync-token.test.ts [hunk 10]
```diff
@@ -1,254 +1,396 @@
...
+    const token = await captureObsidianToken({
+      prompts: scripted.prompts,
+      fetchFn: fetchJsonNonObject(),
+    })
+
+    expect(token).toBeUndefined()
+    expect(scripted.warnings[0]).toBe(
+      "Could not sign in: Unexpected response from Obsidian API (not a JSON object)",
     )
+  })
```
[the remark this answers is not in the document]
Comment (aliasunder) [reply cli/src/__tests__/get-sync-token.test.ts:348]: Already fixed — `fetchJsonNonObject` mock and "returns undefined when the response is valid JSON but not an object" test were added, plus `isJsonObject` rejects arrays.

---
*🔍 ship-check · pr-monitor · Claude Opus 4.6 (1M context)*

File: deploy/railway/README.md [hunk 11]
```diff
@@ -44,24 +44,33 @@ your computer — once, before you click the button.
...
+<summary><strong>Don't have Node.js?</strong></summary>
+
+Install [Docker Desktop](https://www.docker.com/products/docker-desktop/)
+and run the login in a throwaway container instead:
+
+```bash
+docker run --rm -it --entrypoint get-sync-token ghcr.io/aliasunder/vault-cortex:remote
+```
+
+</details>
```
[the remark this answers is not in the document]
Comment (aliasunder) [reply deploy/railway/README.md:73]: Already fixed — both guides now lead with `npx vault-cortex@latest get-sync-token` as the primary path, with Docker in a collapsible "Don't have Node.js?" fallback.

---
*🔍 ship-check · pr-monitor · Claude Opus 4.6 (1M context)*

**Comment by aliasunder** [comment]
The AGENTS.md entry was updated in the first commit of this PR — it already reads:

```
get-sync-token.ts                  # Get-sync-token subcommand (Sync token capture via Obsidian API)
```

The `rootfs/usr/local/bin/get-sync-token` entry (line 59) is a separate file — the shell script inside the container for the manual `docker exec` flow, which is unchanged.

File: cli/src/init.ts [hunk 12]
```diff
@@ -445,25 +447,19 @@ const runRemoteInit = async (
...
-        "Paste the Obsidian Sync token (leave blank to fill in .env later):",
-      )
-    ).trim()
+  // Sign in to Obsidian and capture the Sync token directly via the API.
+  // When the user declines or capture fails, fall back to any token already
+  // in the on-disk .env (a re-init over an existing deployment). Only show
+  // the "run get-sync-token later" guidance when neither source has a token.
+  const capturedToken = await offerSyncTokenCapture(prompts, fetchFn)
+  const existingEnvToken = readEnvObsidianToken(join(targetDir, ".env"))
+  const hasExistingToken = Boolean(capturedToken ?? existingEnvToken)
```
[the remark this answers is not in the document]
Comment (aliasunder) [reply cli/src/init.ts:457]: This test already exists — line 770: "preserves existing token and offers start when capture is declined on re-init". It seeds a `.env` with `OBSIDIAN_AUTH_TOKEN=existing-token`, declines capture (`false`), and asserts `Start the server now?` is asked and `No token yet` is absent from logs.

---
*🔍 ship-check · pr-monitor · Claude Opus 4.6 (1M context)*

File: cli/src/scaffold.ts [hunk 13]
```diff
@@ -151,6 +151,21 @@ export const patchEnvObsidianToken = (
...
+ */
+export const readEnvObsidianToken = (
+  envFilePath: string,
+): string | undefined => {
+  if (!existsSync(envFilePath)) return undefined
+  const match = /^OBSIDIAN_AUTH_TOKEN=(.+)$/m.exec(
+    readFileSync(envFilePath, "utf8"),
+  )
+  return match?.[1].trim() || undefined
+}
```
[the remark this answers is not in the document]
Comment (aliasunder) [reply cli/src/scaffold.ts:167]: These tests already exist — `describe("readEnvObsidianToken")` at scaffold.test.ts:481-521 covers all five cases: missing file, non-empty value, empty value (→ undefined), no line (→ undefined), and whitespace trimming. Added in the same commit as the function.

---
*🔍 ship-check · pr-monitor · Claude Opus 4.6 (1M context)*

File: cli/src/__tests__/init.test.ts [hunk 14]
```diff
@@ -580,35 +577,45 @@ describe("runInit remote flow", () => {
       },
     )
 
-    // No "Generate the token now?" (capture needs Docker) and no "Start the
-    // server now?" (the install warning replaces the start offer).
+    // Token generation is always offered (uses the API, not Docker). With a
+    // blank token (capture declined), no start offer is shown — and since
+    // Docker is not installed, the Docker-not-installed warning fires too.
```
[the remark this answers is not in the document]
Comment (aliasunder) [reply cli/src/__tests__/init.test.ts]: The comment at line 580-581 already reads: "Token generation is always offered (uses the API, not Docker). With a blank token (capture declined), no start offer is shown." — this matches the behavior. The claim about "the Docker-not-installed warning fires too" that the finding references is not present in the current code.

---
*🔍 ship-check · pr-monitor · Claude Opus 4.6 (1M context)*

**Comment by aliasunder** [comment]
**Re: MFA retry timeout handling** — `warnSigninError` is a shared function called by both the initial signin and MFA retry catch blocks (line 151: `warnSigninError(retryError, prompts, true)`). It checks `error.name === "TimeoutError"` first, regardless of the `isMfaRetry` flag, so the MFA retry path already produces "Request timed out — check your internet connection and try again." The bot didn't trace through the shared function.

---
*🔍 ship-check · pr-monitor · Claude Opus 4.6 (1M context)*

File: cli/src/init.ts [hunk 15]
```diff
@@ -75,20 +76,21 @@ const askMode = async (prompts: Prompts): Promise<Mode> => {
...
 /**
- * Offers to auto-capture the Obsidian Sync token via a Docker volume mount.
- * Returns the captured token string, or undefined when the user declines or
- * the capture fails (the caller falls back to a paste prompt).
+ * Offers to sign in to the Obsidian account and capture the Sync token.
+ * Returns the captured token string, or undefined when the user declines
+ * or the capture fails (the caller falls back to a paste prompt).
  */
 const offerSyncTokenCapture = async (
   prompts: Prompts,
```
[the remark this answers is not in the document]
Comment (aliasunder) [reply cli/src/init.ts:85]: Fixed in f2c5bc5 — updated to: "the caller falls back to any token already in the on-disk .env, or shows get-sync-token guidance."

---
*🔍 ship-check · pr-monitor · Claude Opus 4.6 (1M context)*

File: cli/src/scaffold.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (aliasunder) [reply cli/src/scaffold.ts:167]: `readEnvObsidianToken` follows the same first-match pattern as every other scaffold.ts reader — `readEnvPort` (exec), `readEnvVaultPath` (exec), `readEnvPublicUrl` (exec). `readOptionalValue` is in `optional-settings.ts`, a different module with a different concern (it manages user-editable settings where duplicate lines are expected from its own replace-all writes). The CLI-generated `.env` never has duplicate `OBSIDIAN_AUTH_TOKEN` lines; the scenario requires manual hand-editing to create duplicates.

---
*🔍 ship-check · pr-monitor · Claude Opus 4.6 (1M context)*

File: cli/src/__tests__/get-sync-token.test.ts [hunk 16]
```diff
@@ -1,254 +1,396 @@
-import {
-  existsSync,
-  mkdirSync,
-  mkdtempSync,
-  readFileSync,
-  writeFileSync,
-} from "node:fs"
+import { mkdtempSync, readFileSync, writeFileSync } from "node:fs"
```
[the remark this answers is not in the document]
Comment (aliasunder) [reply cli/src/__tests__/get-sync-token.test.ts:1]: This test already exists — line 336: "returns undefined when the response is valid JSON but not an object", using `fetchJsonNonObject()` (line 59). It asserts the exact warning message `"Could not sign in: Unexpected response from Obsidian API (not a JSON object)"`.

---
*🔍 ship-check · pr-monitor · Claude Opus 4.6 (1M context)*

File: cli/src/get-sync-token.ts [hunk 17]
```diff
@@ -1,12 +1,5 @@
-import { existsSync, mkdtempSync, readFileSync, rmSync } from "node:fs"
-import { tmpdir } from "node:os"
 import { join, resolve } from "node:path"
```
[the remark this answers is not in the document]
Comment (aliasunder) [reply cli/src/get-sync-token.ts:1]: Both paths call the shared `warnSigninError` function (line 79), which checks `error.name === "TimeoutError"` first — regardless of whether it is the initial or MFA retry call. The MFA retry catch at line 147 calls `warnSigninError(retryError, prompts, true)`, so a timeout during MFA verification produces "Request timed out — check your internet connection and try again.", not the raw abort message.

The test at line 197 ("shows a timeout message when the MFA retry times out") proves this: it stubs the second fetch call to throw a TimeoutError and asserts the exact timeout warning.

---
*🔍 ship-check · pr-monitor · Claude Opus 4.6 (1M context)*

**Comment by aliasunder** [comment]
**Consolidating umm-actually PR-level comments**

Several PR-level findings duplicate inline threads already resolved with replies, and the remaining new findings are false positives. Addressing each:

**Duplicates of resolved inline threads:**
- "Add a test for declined capture with existing on-disk token" — test exists at line 770 (`preserves existing token and offers start when capture is declined on re-init`). Inline thread replied + resolved.
- "Add unit tests for readEnvObsidianToken edge cases" — 5 tests exist at scaffold.test.ts:481-521. Inline thread replied + resolved.
- "Match docker --env-file's last-wins precedence" — follows established scaffold.ts first-match convention (same as `readEnvPort`, `readEnvVaultPath`, `readEnvPublicUrl`). Inline thread replied + resolved.
- "Correct the misleading comment in the init always-offer test" — comment at line 580 already reads correctly. Inline thread replied + resolved.

**New findings (false positives):**
- "Read OBSIDIAN_AUTH_TOKEN presence from the kept .env on re-init" — the code already does this: `readEnvObsidianToken(join(targetDir, ".env"))` at line 456, combined via `Boolean(capturedToken ?? existingEnvToken)`. The test at line 770 proves the start offer fires when capture is declined but the kept .env has a token.
- "Invert Railway and Render token capture to lead with npx CLI" — the guides already lead with npx (steps 2-3), Docker is in a collapsible "Don't have Node.js?" fallback. The bot noted it compared against prior-review quotes, not the current files.

**Previously addressed:**
- "Update AGENTS.md" — already updated in the first commit of the PR. Replied twice (once from a prior session, once from this monitoring session).
- "Handle timeouts in MFA retry" — `warnSigninError` handles TimeoutError for both paths; test at line 197 proves it. Replied in a prior comment + inline thread.

---
*🔍 ship-check · pr-monitor · Claude Opus 4.6 (1M context)*
